### PR TITLE
Create GetMyDonationDetails action/service, Integration tests, Fixed job, added one column to Donation table (new migration)

### PR DESCRIPTION
**Pull request by badekk**
## Create GetMyDonationDetails

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?

- [ ] Unit test(s)
- [X] Integration test(s)
- [X] Tested manually

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have checked my code and corrected any misspellings
- [] I have added frontend code and checked it for mobile & desktop

### CODE REVIEW

**Comment by mar-pan** [comment]
Task linked: [CU-862j623zq Details](https://app.clickup.com/t/862j623zq)

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            var decodedDonationId = _hashids.DecodeSingleLong(id);
+
+
+            var donationById = _appDbContext.Donations.SingleOrDefaultAsync(c => c.Id == decodedDonationId
+               && c.Status == DbEntityStatus.Active || c.Status == DbEntityStatus.Inactive);
+
+            if (donationById == null)
+            {
+                _logger.LogError("Couldn't find donation in database");
+                throw new ApplicationErrorException("Sorry there seems to be a problem with our service");
```
Comment (h4570) [inline src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:56]: `NotFoundException`

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            }
+
+            var listOfDonationProducts = await _appDbContext.DonationProducts
+                .Include(c => c.Product)
+                .Where(c => c.Donation == donationById.Result
```
Comment (h4570) [inline src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:61]: To jest całkiem do wywalenia, dołożyłeś sobie roboty. Po prostu tam wyżej zrób
```
_appDbContext.Donations
   .Include(c => c.Products)
       .ThenInclude(c => c.Product).SingleOrDefaultAsync(...)
   .Include(c => c.Warehouse)
     .ThenInclude(c => c.Address)
        .ThenInclude(c => c!.Country)
```
i będziesz miał wszystko za jednym strzałem.

BTW.: W `Where()` nie porównuj obiektów. Pamiętaj, że te pseudo kwerendy są tłumaczone na SQL - używaj np `DonationId`.

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            }
+
+            var productResponse = listOfDonationProducts.Select(c => new GetMyDonationDetailsProductResponse
+            {
+                Name = c.Product!.Name,
+                Quantity = c.Quantity,
+                Unit = c.Product.Unit.ToString()
+            }).ToList();
+
+            var warehouseById = await _appDbContext.Warehouses
```
Comment (h4570) [inline src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:77]: Do wywalenia z tego samego powodu co wyżej.

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                _logger.LogError("Couldn't find warehouse in database");
+                throw new ApplicationErrorException("Sorry there seems to be a problem with our service");
+            }
+
+            var response = new GetMyDonationDetailsResponse
+            {
+                Id = _hashids.EncodeLong(donationById.Id),
+                Name = donationById.Result!.Name,
+                CreationDate = donationById.Result.CreationDate,
+                ExpireDate = ExpireDateService.GetExpiredDate4Donation(donationById.Result.CreationDate),
```
Comment (h4570) [inline src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:92]: Ej, to miało być wyliczane przy tworzeniu dotacji, i tutaj tylko `donationById.ExpireDate`

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            {
+                _logger.LogError("Couldn't find any active products in database");
+                throw new ApplicationErrorException("Sorry there seems to be a problem with our service");
+            }
+
+            var productResponse = listOfDonationProducts.Select(c => new GetMyDonationDetailsProductResponse
+            {
+                Name = c.Product!.Name,
+                Quantity = c.Quantity,
+                Unit = c.Product.Unit.ToString()
```
Comment (h4570) [inline src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:74]: `.ToString().ToLowerCase()` z tego co pamiętam to tak prosiłem w zadaniu

File: src/backend/DonateTests/Services/GetMyDonationDetailsActionTest/GetMyDonationDetailsActionTest.cs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                Status = DbEntityStatus.Active
+            };
+            context.Add(donation);
+
+            var donationProduct1 = new DonationProduct
+            {
+                Donation = donation,
+                ProductId = 1,
+                Quantity = 1,
+                UpdateUserId = UserSeeder.ServierUser.Id,
```
Comment (h4570) [inline src/backend/DonateTests/Services/GetMyDonationDetailsActionTest/GetMyDonationDetailsActionTest.cs:112]: `ServierUser` -> `ServiceUser`

File: src/backend/DonateTests/Services/GetMyDonationDetailsActionTest/GetMyDonationDetailsActionTest.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                UpdatedAt = new DateTime(2023, 01, 01, 0, 0, 0, DateTimeKind.Utc),
+                Status = DbEntityStatus.Active
+            };
+            context.Add(donationProduct1);
+
+            var donationProduct2 = new DonationProduct
+            {
+                Donation = donation,
+                ProductId = 2,
```
Comment (h4570) [inline src/backend/DonateTests/Services/GetMyDonationDetailsActionTest/GetMyDonationDetailsActionTest.cs:121]: Bezpieczniej byłoby `ProductId = 2` zmienić na `ProductId = ProductSeeder.Product2.Id`

**Review by h4570 — CHANGES_REQUESTED** [review]

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            _roleService = roleService;
+            _hashids = hashids;
+            _logger = logger;
+        }
+
+        public async Task<GetMyDonationDetailsResponse> GetMyDonationDetails(string id)
+        {
+            var auth0UserInfo = await _currentUserService.GetUserInfo();
+            var currentUser = await _getCurrentUserService.Execute(auth0UserInfo);
+            await _roleService.ThrowIfNoRole(RoleName.Donor, currentUser.Id);
```
[the remark this answers is not in the document]
Comment (badekk) [reply src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:46]: <facePalm> u r right... fixed

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            var decodedDonationId = _hashids.DecodeSingleLong(id);
+
+
+            var donationById = _appDbContext.Donations.SingleOrDefaultAsync(c => c.Id == decodedDonationId
+               && c.Status == DbEntityStatus.Active || c.Status == DbEntityStatus.Inactive);
```
[the remark this answers is not in the document]
Comment (badekk) [reply src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:51]: fixed

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs (hunk 1 above)
Comment (badekk) [reply src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:56]: fixed

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs (hunk 5 above)
Comment (badekk) [reply src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:74]: fixed

File: src/backend/DonateTests/Services/GetMyDonationDetailsActionTest/GetMyDonationDetailsActionTest.cs (hunk 6 above)
Comment (badekk) [reply src/backend/DonateTests/Services/GetMyDonationDetailsActionTest/GetMyDonationDetailsActionTest.cs:112]: done

File: src/backend/DonateTests/Services/GetMyDonationDetailsActionTest/GetMyDonationDetailsActionTest.cs (hunk 7 above)
Comment (badekk) [reply src/backend/DonateTests/Services/GetMyDonationDetailsActionTest/GetMyDonationDetailsActionTest.cs:121]: nice one, resolved

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs (hunk 2 above)
Comment (badekk) [reply src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:61]: Gucci geng

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs (hunk 3 above)
Comment (badekk) [reply src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs:77]: gucci geng2

File: src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs (hunk 4 above)
Comment (badekk) [reply src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs]: fixed

**Review by h4570 — APPROVED** [review]

**Pull request merged**